src/backend/distributed/utils/enable_ssl.c

```diff
@@ -0,0 +1,429 @@
+ *
+ * enable_ssl.c
+ *    UDF and Utilities for enabling ssl during citus setup
+ *
+ * Copyright (c) 2018, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "distributed/connection_management.h"
+#include "distributed/worker_protocol.h"
+#include "fmgr.h"
+#include "libpq/libpq.h"
+#include "nodes/parsenodes.h"
+#include "postmaster/postmaster.h"
+#include "utils/guc.h"
+
+#ifdef USE_OPENSSL
+#include "openssl/dsa.h"
+#include "openssl/err.h"
+#include "openssl/pem.h"
+#include "openssl/rsa.h"
+#include "openssl/ssl.h"
+#include "openssl/x509.h"
+#endif
+
+
+#define DirectFunctionCall0(func) \
+	DirectFunctionCall0Coll(func, InvalidOid)
+
+#define ENABLE_SSL_QUERY "ALTER SYSTEM SET ssl TO on;"
+#define RESET_CITUS_NODE_CONNINFO \
+	"ALTER SYSTEM SET citus.node_conninfo TO 'sslmode=prefer';"
+
+#define CITUS_AUTO_SSL_COMMON_NAME "citus-auto-ssl"
+#define X509_SUBJECT_COMMON_NAME "CN"
+
+
+static bool ShouldUseAutoSSL(void);
+static Datum DirectFunctionCall0Coll(PGFunction func, Oid collation);
+static void EnsureFreeOpenSSLResource(MemoryContextCallbackFunction free, void *arg);
+
+/* use pg's implementation that is not exposed in a header file */
+extern Datum pg_reload_conf(PG_FUNCTION_ARGS);
+
+
+#ifdef USE_SSL
+static bool CreateCertificatesWhenNeeded(void);
+static EVP_PKEY * generate_key_rsa(void);
+static X509 * generate_x509(EVP_PKEY *pkey);
+static bool StoreCertificate(EVP_PKEY *pkey, X509 *x509);
+#endif /* USE_SSL */
+
+
+PG_FUNCTION_INFO_V1(citus_setup_ssl);
+PG_FUNCTION_INFO_V1(citus_reset_default_for_node_conninfo);
+
+
+/*
+ * citus_setup_ssl is called during the first creation of a citus extension. It configures
+ * postgres to use ssl if not already on. During this process it will create certificates
+ * if they are not already installed in the configured location.
+ */
+Datum
+citus_setup_ssl(PG_FUNCTION_ARGS)
+{
+#ifndef USE_SSL
+	ereport(WARNING, (errmsg("can not setup ssl on postgres that is not compiled with "
+							 "ssl support")));
+#else /* USE_SSL */
+	if (!EnableSSL && ShouldUseAutoSSL())
+	{
+		Node *enableSSLParseTree = NULL;
+
+		elog(LOG, "citus extension created on postgres without ssl, turning it on");
+
+		/* execute the alter system statement to enable ssl on within postgres */
+		enableSSLParseTree = ParseTreeNode(ENABLE_SSL_QUERY);
+		AlterSystemSetConfigFile((AlterSystemStmt *) enableSSLParseTree);
+
+		/*
+		 * ssl=on requires that a key and certificate are present, since we have
+		 * enabled ssl mode here chances are the user didn't install any credentials.
+		 * This function will check if they are available and if not it will generate
+		 * self singed certificates.
+		 */
+		CreateCertificatesWhenNeeded();
+
+#if PG_VERSION_NUM >= 100000
+
+		/* changing ssl configuration requires a reload of the configuration */
+		DirectFunctionCall0(pg_reload_conf);
+		ProcessConfigFile(PGC_SIGHUP);
```

Comment:
add extra comment why we need to reload the configuration like this besides the pg_reload_conf